src/batch/src/executor2/filter.rs

```diff
@@ -12,104 +12,88 @@
-            } else {
-                let child_input = self.fetch_one_chunk().await?;
-                if let Some(data_chunk) = child_input {
-                    self.last_input = Some(SlicedDataChunk::new_checked(data_chunk)?);
-                } else {
-                    // We should return here since nothing come from child.
-                    return if let Some(left) = self.chunk_builder.consume_all()? {
-                        Ok(Some(left))
-                    } else {
-                        Ok(None)
-                    };
-                }
-            }
-        }
-    }
-
-    async fn close(&mut self) -> Result<()> {
-        self.child.close().await
-    }
-
+impl Executor2 for FilterExecutor2 {
     fn schema(&self) -> &Schema {
         self.child.schema()
     }
 
     fn identity(&self) -> &str {
         &self.identity
     }
+
+    fn execute(self: Box<Self>) -> BoxedDataChunkStream {
+        self.do_execute()
+    }
 }
 
-impl FilterExecutor {
-    /// Fetch one chunk from child.
-    async fn fetch_one_chunk(&mut self) -> Result<Option<DataChunk>> {
-        if self.child_can_be_nexted {
-            if let Some(data_chunk) = self.child.next().await? {
-                let data_chunk = data_chunk.compact()?;
-                let vis_array = self.expr.eval(&data_chunk)?;
-                return if let Bool(vis) = vis_array.as_ref() {
-                    let vis = vis.try_into()?;
-                    let data_chunk = data_chunk.with_visibility(vis);
-                    Ok(Some(data_chunk))
-                } else {
-                    Err(InternalError("Filter can only receive bool array".to_string()).into())
-                };
+impl FilterExecutor2 {
+    #[try_stream(boxed, ok = DataChunk, error = RwError)]
+    async fn do_execute(self: Box<Self>) {
+        let mut data_chunk_builder =
+            DataChunkBuilder::new_with_default_size(self.child.schema().data_types());
+
+        #[for_await]
+        for data_chunk in self.child.execute() {
+            let data_chunk = data_chunk?.compact()?;
+            let vis_array = self.expr.eval(&data_chunk)?;
+
+            if let Bool(vis) = vis_array.as_ref() {
+                let mut sliced_data_chunk =
+                    SlicedDataChunk::new_checked(data_chunk.with_visibility(vis.try_into()?))?;
+
+                loop {
+                    let (left_data, output) = data_chunk_builder.append_chunk(sliced_data_chunk)?;
+                    match (left_data, output) {
+                        (Some(left_data), Some(output)) => {
+                            sliced_data_chunk = left_data;
+                            yield output;
+                        }
+                        (None, Some(output)) => {
+                            yield output;
+                            break;
+                        }
+                        (None, None) => {
+                            break;
+                        }
+                        _ => {
```

Comment:
A question may not relates to this PR: Why streaming executor has no `DataChunkBuilder` but batch have?

Then we may have very similar logic in each batch executor doing `append_chunk` stuffs.

Reply:
I've took a look at streaming executor yet, but this may be because streaming executor haven't consider splitting data chunk yet.

Reply:
> I've took a look at streaming executor yet, but this may be because streaming executor haven't consider splitting data chunk yet.

Yes it did not. The problem of `DataChunkBuilder` is it have too much code-invasion. I'm not sure whether we needs better way to do this.

Reply:
Welcome to raise proposals.

Reply:
> The problem of DataChunkBuilder is it have too much code-invasion

+1. `DataChunkBuilder` may have better implementation